docs/cookbook/rag.md

```diff
@@ -0,0 +1,257 @@
+
+Our dataset will be a collection of news articles from 2004 pulled from this [public Github database.](https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv) We will pick a topic, then ask our chat model to give a summary of 4 relevant articles.
+
+### Dataset/Preprocessing
+
+Chat and embedding models have token limits on a single call - this wasn't an issue in the soccer example, but these constraints come into play with much larger texts and datasets. For the purposes of this recipe, we split the large texts into roughly equal sizes using helper function `split_text()`. Since we want to be able to fit multiple articles into our final prompt, let's set `max_tokens=1000` and split the text into how many thousands of tokens the original text comprises of.
+
+```python
+def split_text(text: str, tokens: list[int], max_tokens: int) -> list[str]:
+    """Roughly splits a text into chunks according to max_tokens.
+
+    Text is split into equal word counts, with number of splits determined by how many
+    times `max_tokens` goes into the total number of tokens (including partially). Note
+    that tokens and characters do not have an exact correspondence, so in certain edge
+    cases a chunk may be slightly larger than max_tokens.
+
+    Args:
+        text: The text to split.
+        tokens: How many tokens `text` is.
+        max_tokens: The (rough) number of maximum tokens per chunk.
+
+    Returns:
+        A list of the split texts.
+    """
+    words = text.split()
+    num_splits = len(tokens) // max_tokens + 1
+    split_texts = []
+
+    for i in range(num_splits):
+        start = i * len(words) // num_splits
+        end = (i + 1) * len(words) // num_splits
+        split_texts.append(" ".join(words[start:end]))
+
+    return split_texts
+```
+
+```python
+MODEL = "gpt-3.5-turbo"
+TEXT_COLUMN = "Text"
+URL = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+
+df = load_data(URL)
+
+encoder = tiktoken.encoding_for_model(MODEL)
+split_articles = []
+
+for i, row in df.iterrows():
+    text = row[TEXT_COLUMN]
+    tokens = encoder.encode(text)
+    max_tokens = 1000
+    
+    if len(tokens) > max_tokens:
+        split_articles += split_text(text, tokens, max_tokens)
+        df.drop(i, inplace=True)
+
+    # Long texts which were dropped from the dataframe are now readded.
+    df = pd.concat(
+        [df, pd.DataFrame(split_articles, columns=[TEXT_COLUMN])], ignore_index=True
+    )
+```
+
+### Batched Embeddings
+
+As before, we embed our texts. However, we can speed up the process and minimize token usage by embedding in batches. Since `max_tokens=1000` and the model `text-embedding-ada-002` has a token limit of 8096, we can pass in batches of 8.
+```python
+chat = OpenAIChat(model=MODEL)
+
+EMBEDDINGS_COLUMN = "embeddings"
+
+batch_size = 8    
```

Comment:
inconsistent use of upper snake and lower snake case, applies to all other variables